### DISCUSSION TITLE
Multiple Identifiers for an http label

### DISCUSSION THREAD

**Discussion by DaneLyttinen** [awslabs/smithy#1967] [Q&A]
I have tried looking into the docs to solve this but maybe I have overlooked it but it would be great if someone could clarify how to handles this scenario. 

If we have a resource with two or more identifiers like the one below.

```smithy
resource AwesomeResource{
    identifiers:{id: String, id2 : String}
 }
```

And then want to define a Get operation

```smithy
@http(method: "GET", uri:"/identities/{id}", code:200)
operation GetAwesomeResource {
    input: GetAwesomeResourceInput
    output: GetAwesomeResourceOutput
}
```

Alongside with the input

```smithy
@input
structure GetIdentityInput{
    @required
    @httpLabel
    id: String
}
```

The problem starts becoming a bit apparent, how do we handle using the same label (in this case id) for both identifiers (id & id2)? I have tried quite a few different ways such as using the greedy label but that cannot be used twice to point to different identifiers it seems. Don't believe we can either define a new operation which is identical except referencing id2 and then using mixins would work? But either way I've struggled a bit with this and was wondering if there is some sort of "or" operation that could be used or something along those lines. If someone could clarify how to handle this properly that'd be great.

**Comment by JordonPhillips** [comment]
I think there might be some confusion about what it means for a resource to have multiple identifiers. It isn't that the resource has two separate identifiers that can each uniquely identify the resource. Rather, it means that the resource can only be uniquely identified if all of its identifier shapes are filled out. Consider the following example:

```smithy
resource ObjectResource {
    identifiers: {
        bucket: String
        path: String
    }
}
```

Both `bucket` and `path` are required to identify the resource. Since you can't uniquely identify the resource without both, you can't have a `get` operation without both. You would end up with a uri like `/object/{bucket}/{path}`. Additionally, any member with the `@httpLabel` trait must also have the `@required` trait, so if you want them bound to the path you have set either of them anyway.

With that out of the way, if you want to have an operation where either can be missing, you just need to bind them somewhere else. If you want them in the URI, then binding them to the path is an option, like so:

```smithy
@readonly
@http(method: "GET", uri: "/objects", code: 200)
operation SearchObjects {
    input := {
        @httpQuery("bucket")
        bucket: String

        @httpQuery("path")
        path: String
    }
    output := {
        objects: ObjectList
    }
}

list ObjectResource {
    member: Object
}
```

This is called a [collection operation](https://smithy.io/2.0/spec/service-types.html#collection-operations) because rather than operating on a single resource, you're operating on the collection of all instances of that resource. Hence, the uri is `/objects` rather than `/object/{bucket}/{path}` and the return type is a list since multiple instances could match. At runtime you'd end up with urls like `/objects?bucket=foo` or `/objects?path=bar`.
